terraform/eval_apply.go

```diff
@@ -561,8 +561,15 @@ func (n *EvalApplyProvisioners) apply(ctx EvalContext, provs []*configs.Provisio
 		provisioner := ctx.Provisioner(prov.Type)
 		schema := ctx.ProvisionerSchema(prov.Type)
 
-		forEach, forEachDiags := evaluateResourceForEachExpression(n.ResourceConfig.ForEach, ctx)
-		diags = diags.Append(forEachDiags)
+		var forEach map[string]cty.Value
+
+		// We can't evaluate the for_each expression during a destroy
```

Comment:
The logic here seems fine, but I feel like I want the commentary around this to be a little clearer about the intent because right now this use of `forEach` without populating it first has the pattern of a common bug I look for in code review. Perhaps we could add a couple extra comments to help be more explicit for future-us who don't remember about this PR...

For the `EvalKeyForInstanceKey` call below, perhaps:

```go
// For a destroy-time provisioner forEach is intentionally nil here, which
// EvalDataForInstanceKey responds to by not populating EachValue
// in its result. That's okay because each.value is prohibited for
// destroy-time provisioners.
```

...and then in the docstring for `EvalDataForInstanceKey` itself, perhaps:

```go
// EvalDataForInstanceKey constructs a suitable InstanceKeyEvalData for
// evaluating in a context that has the given instance key.
//
// The forEachMap argument can be nil when preparing for evaluation
// in a context where each.value is prohibited, such as a destroy-time
// provisioner. In that case, the returned EachValue will always be
// cty.NilVal.
```

Reply:
Those are great! added verbatim.